### PR TITLE
add thin_provisioning cases according to some bugs

### PR DESCRIPTION
**Pull request by PyLearner**
ID: 1354233, 1373048

### CODE REVIEW

**Comment by suqinhuang** [comment]
commit log pls

**Comment by suqinhuang** [comment]
could you pls update your summary and commit log

**Comment by PyLearner** [comment]
this patch is about 2 cases.
1. guest should always show unmap mode.
2. image convert should use less time than old version qemu.

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by Xueqiang Wei <xuwei@redhat.com>

**Comment by suqinhuang** [comment]
id pls

File: qemu/tests/cfg/thin_provisioning_other.cfg [hunk 1]
```diff
@@ -0,0 +1,28 @@
+- thin_provisioning_other:
```
Comment (suqinhuang) [inline qemu/tests/cfg/thin_provisioning_other.cfg]: not a good name, generally people can get the main info from the name

File: qemu/tests/thin_provisioning_guest_show_unmap.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+from autotest.client import os_dep
+from avocado.core import exceptions
+from virttest import env_process
+from qemu.tests import thin_provisioning
+
+
+@error.context_aware
+def run(test, params, env):
+    """
+      'thin-provisioning' functions of fixed bugs
```
Comment (suqinhuang) [inline qemu/tests/thin_provisioning_guest_show_unmap.py]: what does the bugs fixed? you should descript the function

File: qemu/tests/cfg/thin_provisioning_other.cfg [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    image_format_image_other = raw
+    force_create_image_image_other = no
+    image_raw_device_image_other = yes
+    drive_format_image_other = scsi-hd
+    drv_extra_params_image_other = "discard=on"
+    guest_scsi_disk_cmd = "fdisk -l 2>&1 | grep "Disk /dev/[vs]d[a-z]" | cut -d" " -f2 | cut -d: -f1 | tail -1"
+
+    variants:
+        - always_show_unmap:
+            pre_command = "modprobe scsi_debug lbpu=1 lbpws=1 lbprz=0"
```
Comment (suqinhuang) [inline qemu/tests/cfg/thin_provisioning_other.cfg]: where will pre_command be run, didn't find it in your code

File: qemu/tests/thin_provisioning_guest_show_unmap.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    # Destroy all vms to avoid emulated disk marked drity before start test
+    for vm in env.get_all_vms():
+        if vm:
+            vm.destroy()
+            env.unregister_vm(vm.name)
+
+    os_dep.command("lsscsi")
+    host_id, disk_name = thin_provisioning.get_scsi_disk(session=None)
+    provisioning_mode = thin_provisioning.get_provisioning_mode(host_id, disk_name)
+    error.context("TEST STEP 1: The provisioning_mode is {:s}".format(provisioning_mode), logging.info)
```
Comment (suqinhuang) [inline qemu/tests/thin_provisioning_guest_show_unmap.py]: step is what you will do, not the result. you should add the step description before "lsscsi"

File: qemu/tests/thin_provisioning_guest_show_unmap.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      :param env: Dictionary with test environment.
+      """
+
+    # Destroy all vms to avoid emulated disk marked drity before start test
+    for vm in env.get_all_vms():
+        if vm:
+            vm.destroy()
+            env.unregister_vm(vm.name)
+
+    os_dep.command("lsscsi")
```
Comment (suqinhuang) [inline qemu/tests/thin_provisioning_guest_show_unmap.py]: where did you analysis the output?

File: qemu/tests/cfg/thin_provisioning_other.cfg (hunk 3 above)
Comment (PyLearner) [inline qemu/tests/cfg/thin_provisioning_other.cfg]: it is used in the preprocess stage

File: qemu/tests/thin_provisioning_guest_show_unmap.py (hunk 5 above)
Comment (PyLearner) [inline qemu/tests/thin_provisioning_guest_show_unmap.py]: hi, it is analysied in the thin_provisioning.py , function: get_scsi_disk(session=None)

**Comment by PyLearner** [comment]
hi @suqinhuang  @XueqiangWei  @CongLi  @pingl2016  @aliang123  @ngu-niny  , could you help to review, thanks.

**Comment by CongLi** [comment]
Hi @PyLearner 

Could you please update your commit log?
"this patch is about 2 cases. " is really strange.
Maybe "Add 2 new cases about thin_provisioning.\n 1. *** 2. ***" is better.

**Comment by PyLearner** [comment]
@CongLi , updated, could you help to review , thanks.

**Comment by PyLearner** [comment]
@suqinhuang , updated, could you help to review ? thanks

File: qemu/tests/thin_provisioning_optimize.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import logging
```
Comment (xutian) [inline qemu/tests/thin_provisioning_optimize.py]: take care order please!!

File: qemu/tests/thin_provisioning_optimize.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        qemu_img_binary, base_dir, base_dir)
+    process.system(image_create_cmd, shell=True)
+    output = process.system_output(image_convert_cmd, shell=True)
+    realtime = re.search(r"real\s+\dm(.*)s", output)
+    if realtime is None:
+        raise exceptions.TestError(
+            "Faild to get the realtime from {}".format(output))
+    realtime = float(realtime.group(1))
+    logging.info("real time is : {:f}".format(realtime))
+    if realtime >= standard_time:
```
Comment (xutian) [inline qemu/tests/thin_provisioning_optimize.py]: where you get stand time?  image create time depends on disk where your image keep or image_size and image_format and so on...  so, how to ensure your test is right.

Comment (PyLearner) [inline qemu/tests/thin_provisioning_optimize.py]: hi @xutian , thanks for your comment , the standard time i set is according to the bug  can check it in the bug id.
before the bug is fixed ,the time used is around 0.4
after the bug is fixed, the time used is like 0.02
so when i try to automate it , i set the standard time to it .

could you give some advice about how to automate this case ,thanks a lot.

**Comment by xutian** [comment]
Invalid ID format, please modify it like ID: id1 id2 ...

**Comment by xutian** [comment]
I don't think it's a reasonable case. wirtesame and unmap just to scsi command, if we set discard=on, qemu will issue unmap or writesame command to backend block device to ask device release delete file space. I don't know any help for raw/qcow2 image convert.

**Pull request closed without merging**